src/inmanta/server/services/orchestrationservice.py

```diff
@@ -384,6 +384,39 @@ async def _purge_versions(self) -> None:
                 for v in delete_list:
                     await version_dict[v].delete_cascade()
 
+        # Cleanup old agents from agent table in db
+        async with data.Agent.get_connection() as con:
+            await con.execute(
+                """
+-- Delete all records from the agent table where the (environment, name) pair
+-- is not in the set of (environment, name) pairs returned by the subquery.
+DELETE FROM public.agent
+WHERE (environment, name) NOT IN (
+    SELECT DISTINCT environment_id as environment, agent as name
+    FROM (
+        -- agent is in the agent map
+        SELECT e.id as environment_id, map.key as agent
+        FROM environment e
+        CROSS JOIN LATERAL jsonb_each(e.settings->'autostart_agent_map') AS map(key, value)
+
+        -- Union with the set of (environment, name) pairs for all agents that
+        -- have a primary ID set (that are not down)
+        UNION ALL
```

Comment:
I thought that was how you could join the results of multiple queries but it seems, just UNION is better here as it will remove the duplicates already

Reply:
I agree.